File: centwise/apps/ios/Centwise/Data/Fakes/ReviewQueueRepository.swift
import Foundation
import Combine

public struct ReviewQueueItem: Identifiable, Equatable {
    public let id: String
    public let sender: String
    public let rawSms: String
    public let timestamp: Date
    public let candidateAmount: Double?
    public let candidateParty: String?
    public let candidateType: TransactionType?
    public let reason: String

    public init(
        id: String = UUID().uuidString,
        sender: String,
        rawSms: String,
        timestamp: Date = Date(),
        candidateAmount: Double? = nil,
        candidateParty: String? = nil,
        candidateType: TransactionType? = nil,
        reason: String = "Format needs confirmation"
    ) {
        self.id = id
        self.sender = sender
        self.rawSms = rawSms
        self.timestamp = timestamp
        self.candidateAmount = candidateAmount
        self.candidateParty = candidateParty
        self.candidateType = candidateType
        self.reason = reason
    }
}

public final class ReviewQueueRepository: ObservableObject {
    public static let shared = ReviewQueueRepository()

    @Published public private(set) var items: [ReviewQueueItem] = []

    public init() {}

    public func addItem(_ item: ReviewQueueItem) {
        if !items.contains(where: { $0.rawSms == item.rawSms }) {
            items.insert(item, at: 0)
        }
    }

    public func dismissItem(id: String) {
        items.removeAll { $0.id == id }
    }

    public func confirmAsTransaction(item: ReviewQueueItem, transaction: CentwiseTransaction) {
        FakeTransactionRepository.shared.addTransaction(transaction)
        dismissItem(id: item.id)
    }
}

File: apps/android/app/src/main/kotlin/com/centwise/data/repository/ReviewQueueRepository.kt
package com.centwise.data.repository

import com.centwise.data.models.ReviewQueueItem
import com.centwise.data.models.TransactionItem
import com.centwise.core.backend.CentwiseRustBackend
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map

/**
 * Repository managing unparsed or ambiguous SMS messages waiting in the review queue.
 */
class ReviewQueueRepository private constructor() {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val _items = MutableStateFlow<List<ReviewQueueItem>>(emptyList())
    val items: StateFlow<List<ReviewQueueItem>> = _items.asStateFlow()

    val pendingCount = _items.map { it.size }

    fun refresh() {
        scope.launch {
            _items.value = CentwiseRustBackend.listReviewQueue().map { item ->
                ReviewQueueItem(
                    id = item.id,
                    sender = item.sender ?: "Financial SMS",
                    rawSms = item.rawSms,
                    timestamp = item.receivedAtEpochMs,
                    candidateAmount = item.candidateAmountMinor?.div(100.0),
                    candidateParty = item.party ?: item.merchant,
                    candidateType = item.candidateKind?.let { kind ->
                        when (kind) {
                            com.centwise.core.uniffi.TransactionKind.EXPENSE -> com.centwise.data.models.TransactionType.EXPENSE
                            com.centwise.core.uniffi.TransactionKind.INCOME -> com.centwise.data.models.TransactionType.INCOME
                            com.centwise.core.uniffi.TransactionKind.TRANSFER -> com.centwise.data.models.TransactionType.TRANSFER
                            com.centwise.core.uniffi.TransactionKind.REFUND -> com.centwise.data.models.TransactionType.CREDIT
                        }
                    },
                    reference = item.reference,
                    reason = item.reason
                )
            }
        }
    }

    fun dismissItem(id: String) {
        scope.launch {
            if (CentwiseRustBackend.dismissReviewQueueItem(id)) refresh()
        }
    }

    suspend fun confirmAsTransaction(item: ReviewQueueItem, transaction: TransactionItem): Boolean =
        withContext(Dispatchers.IO) {
            if (!CentwiseRustBackend.convertReviewQueueItem(item.id, transaction)) return@withContext false
            refresh()
            TransactionRepository.shared.loadFromRust()
            true
        }

    companion object {
        val shared = ReviewQueueRepository()
    }
}
